pkg/clustermesh/mcsapi/metrics.go

```diff
@@ -0,0 +1,96 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package mcsapi
+
+import (
+	"context"
+	"log/slog"
+
+	"github.com/prometheus/client_golang/prometheus"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	operatorMetrics "github.com/cilium/cilium/operator/metrics"
+	k8sClient "github.com/cilium/cilium/pkg/k8s/client"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/metrics"
+)
+
+func registerMCSAPICollector(logger *slog.Logger, client k8sClient.Clientset) {
+	operatorMetrics.Registry.MustRegister(&mcsAPICollector{
+		logger: logger,
+		client: client,
+
+		serviceExportCreated: prometheus.NewDesc(
+			prometheus.BuildFQName(metrics.CiliumOperatorNamespace, "", "service_export_created"),
+			"Number of ServiceExport created in the local cluster",
+			[]string{"serviceexport", "namespace"}, nil),
```

Comment:
These metrics have an unbounded number of label tuples (implied by the namespace/name pair), which can put a significant pressure on prometheus, especially in case of large scale clusters. Is such granularity really necessary? Aggregating per namespace would help reducing cardinality, and it may be enough to be acceptable considering that these are exposed by the Cilium operator only.

/cc @cilium/metrics @marseel for opinions.

Reply:
I was wondering about that too but for this initial version after giving out some thought I started from what kube-state-metrics is doing but I could revisit that

Reply:
Also technically this could be done directly in kube-state-metrics but apparently they won't support non native resource, there was a similar point for GW-API here: https://github.com/kubernetes/kube-state-metrics/pull/2452#issuecomment-2532745890

Reply:
While checking more kube-state-metrics doc I noticed this https://github.com/kubernetes/kube-state-metrics/blob/main/docs/design/metrics-best-practices.md#avoid-pre-computation which might be relevant for this discussion

Reply:
I guess that's mainly a tradeoff between expressiveness and cardinality. AFAIK, kube-state-metric has always been in favor of very granular metrics, which can be reasonable in that context, given that it is a single instance component, opt-in and it can be tuned to skip certain metrics if necessary. OTOH, some of the Cilium metrics had/have significant cardinality problems, exacerbated by the fact that Cilium agents run as a daemonset, so the metrics cardinality grows with the number of nodes, easily resulting in millions of series in very large clusters, with the consequent struggles. That said, I agree that these specific metrics are less concerning from that point of view, given that only exposed by the Cilium operator.